### PR TITLE
Enable ratchet tree extensions for MlsGroup

### PR DESCRIPTION
**Pull request by raphaelrobert**
This PR enables the ratchet tree extensions for `MlsGroup` and provides a positive and negative test.

(this PR initially would also add the extension to the group context, which was wrong)

### CODE REVIEW

File: src/group/mls_group/new_from_welcome.rs
Comment (kkohbrok) [inline src/group/mls_group/new_from_welcome.rs:91]: Old code?

File: src/group/mls_group/new_from_welcome.rs
Comment (kkohbrok) [inline src/group/mls_group/new_from_welcome.rs:82]: We're iterating twice over the extensions here. I think you can optimize by filtering first, then checking for duplicates and then using the one that's left (if there is one).

File: src/extensions/test_extensions.rs
Comment (kkohbrok) [inline src/extensions/test_extensions.rs:67]: I think this can be replaced by the `ctest` macro.

**Review by kkohbrok — CHANGES_REQUESTED** [review]
Thanks for tackling this! Just a few comments.

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
What Konrad said.
And please move all the unrelated formatting changes to a separate PR.

File: src/group/mls_group/create_commit.rs
Comment (raphaelrobert) [reply src/group/mls_group/create_commit.rs:103]: I agree that would be a good optimisation for larger groups. Ultimately you need to clone the whole tree of course when you send out a Welcome message, but you shouldn't have to do it when you just apply a Commit.

File: src/group/mls_group/new_from_welcome.rs
Comment (raphaelrobert) [reply src/group/mls_group/new_from_welcome.rs:82]: Where else are we iterating? I'm not really concerned about performance here, the list of extensions should be < 10 most of the time.

File: src/group/mls_group/create_commit.rs
Comment (raphaelrobert) [reply src/group/mls_group/create_commit.rs:103]: I looked into this again and I don't see an easy way to improve the performance here. The problem is that we need the full payload of the group extensions for the group context, which is just a serialisation of the payload. The byte vector is later used to derive the epoch secret. So in order to get rid of all duplication of the public tree, the epoch secret would have to be generated by iterating over the tree. Looking into that, it turns out that it leads to a number of issues:
 - serde cannot deserialise generic impls, meaning we would have tor write all deserialisation for all extensions by hand
 - we need to introduce lifetimes for all extensions (because this particular one now depends on the lifetime of a tree), which is complicated at best and potentially impossible

Given the above, I'd like to punt on making this more efficient for now. This only introduces an overhead when the ratchet tree extension is actually used, which by design is is incompatible with really large groups anyway.

Comment (kkohbrok) [reply src/group/mls_group/create_commit.rs:103]: Thanks for looking into this! I see that it's complicated and I think we can just land this for now. I feel like we've now run into a couple of these issues that we likely want to revisit later to improve performance, so I've created #359 for this issue in particular and #358 to keep track of existing performance bottlenecks.

**Review by kkohbrok — APPROVED** [review]
Just the minor issue with the redundant iteration left. Otherwise this looks really good!

**Review by franziskuskiefer — COMMENTED** [review]
looks like @kkohbrok has this covered

**Review by franziskuskiefer — APPROVED** [review]
but looks like I still need to approve...

**Comment by raphaelrobert** [comment]
There's a pending discussion about what should be included in the `GroupContext`.  The consensus seems to be that the ratchet tree extension should not be included, so I need to rewrite this.

**Comment by raphaelrobert** [comment]
I now refactored the PR so that the ratchet tree extension is not added to the group context. Since the group extensions are not clearly defined in the spec, #186 remains open for now.

**Comment by kkohbrok** [comment]
If the RatchetTreeExtension is not a group extension that members agree on, we could in theory handle it on a per-client basis. Right now, it's "inherited" by clients in way. If they receive a Welcome with a RatchetTreeExtension, they will in the future also include a RatchetTreeExtension in any Welcome that they send. Should we maybe move this decision to the ManagedGroupConfig?

File: src/extensions/ratchet_tree_extension.rs
Comment (franziskuskiefer) [inline src/extensions/ratchet_tree_extension.rs:38]: Shall we leave this in then until the issue and https://github.com/mlswg/mls-protocol/pull/464 are resolved?

File: src/group/mls_group/new_from_welcome.rs
Comment (franziskuskiefer) [inline src/group/mls_group/new_from_welcome.rs:87]: Not sure I understand why they need cloning. But this is a performance issue that will either become one later and we can fix it then, or it doesn't 🙂

File: src/group/mls_group/new_from_welcome.rs
Comment (franziskuskiefer) [inline src/group/mls_group/new_from_welcome.rs:92]:
```suggestion
            // Throw an error if there is more than one ratchet tree extension.
```

Comment (franziskuskiefer) [reply src/group/mls_group/new_from_welcome.rs:92]: Do we have a test for this?

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
This might be something that's not clearly defined in the spec. But we can still take an implementation decision here that would still be fine with the spec. But I'm ok with leaving it like this and seeing what comes from https://github.com/mlswg/mls-protocol/pull/464

lgtm but we should see if we can test the more than one ratchet tree extension error case.

File: src/extensions/ratchet_tree_extension.rs
Comment (raphaelrobert) [reply src/extensions/ratchet_tree_extension.rs:38]: Yes, #186 is essentially blocked by that.

File: src/group/mls_group/new_from_welcome.rs
Comment (raphaelrobert) [reply src/group/mls_group/new_from_welcome.rs:87]: We need an actual copy here for the Welcome message.

**Comment by raphaelrobert** [comment]
> If the RatchetTreeExtension is not a group extension that members agree on, we could in theory handle it on a per-client basis. Right now, it's "inherited" by clients in way. If they receive a Welcome with a RatchetTreeExtension, they will in the future also include a RatchetTreeExtension in any Welcome that they send. Should we maybe move this decision to the ManagedGroupConfig?

This is definitely an implementation decision we take here. We still need to implement it for `ManagedGroup` though.

**Review by franziskuskiefer — APPROVED** [review]
lgtm

**Pull request merged**
